pkg/networkservice/ipam/strictipam/server.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2024 Cisco and its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package strictipam provides a networkservice.NetworkService Server chain element for building an IPAM server that prevents IP context configuration out of the settings scope
+package strictipam
+
+import (
+	"context"
+	"net"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/ippool"
+)
+
+type strictIPAMServer struct {
+	ipPool *ippool.IPPool
+}
+
+// NewServer - returns a new ipam networkservice.NetworkServiceServer that validates the incoming IP context parameters and resets them based on the validation result.
+func NewServer(newIPAMServer func(...*net.IPNet) networkservice.NetworkServiceServer, prefixes ...*net.IPNet) networkservice.NetworkServiceServer {
+	if newIPAMServer == nil {
+		panic("newIPAMServer should not be nil")
+	}
+	var ipPool = ippool.New(net.IPv6len)
+	for _, p := range prefixes {
+		ipPool.AddNet(p)
+	}
+	return next.NewNetworkServiceServer(
+		&strictIPAMServer{ipPool: ipPool},
+		newIPAMServer(prefixes...),
+	)
+}
+
+func (n *strictIPAMServer) areAddressesValid(addresses []string) bool {
```

Comment:
It depends on each NSE and IPAM needs, but here nothing prevent a NSC to request the connection with an IP that is used by another connection.

Reply:
Yes, it's right. With this patch, we're shifting the responsibility of managing IP contexts to NSEs. I also added a new IPAM wrapper chain element that can be used for clearing unepected addresses (if someone will need this functionality in the future). At this moment, it's used only in unit tests, but it can also be used for common cases, for example, as described in the original issue: https://github.com/networkservicemesh/deployments-k8s/issues/9888 ;)